homeassistant/components/growatt_server/__init__.py

```diff
@@ -1,21 +1,98 @@
 """The Growatt server PV inverter sensor integration."""
 
-from homeassistant import config_entries
+import growattServer
+
 from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_URL, CONF_USERNAME
 from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+
+from .const import (
+    CONF_PLANT_ID,
+    DEFAULT_PLANT_ID,
+    DEFAULT_URL,
+    DEPRECATED_URLS,
+    LOGIN_INVALID_AUTH_CODE,
+    PLATFORMS,
+)
+from .coordinator import GrowattCoordinator
+from .models import GrowattRuntimeData
+
+
+def get_device_list(
+    api: growattServer.GrowattApi, config: dict[str, str]
+) -> tuple[list[dict[str, str]], str]:
+    """Retrieve the device list for the selected plant."""
+    plant_id = config[CONF_PLANT_ID]
+
+    # Log in to api and fetch first plant if no plant id is defined.
+    login_response = api.login(config[CONF_USERNAME], config[CONF_PASSWORD])
+    if (
+        not login_response["success"]
+        and login_response["msg"] == LOGIN_INVALID_AUTH_CODE
+    ):
+        raise ConfigEntryError("Username, Password or URL may be incorrect!")
+    user_id = login_response["user"]["id"]
+    if plant_id == DEFAULT_PLANT_ID:
+        plant_info = api.plant_list(user_id)
+        plant_id = plant_info["data"][0]["plantId"]
+
+    # Get a list of devices for specified plant to add sensors for.
+    devices = api.device_list(plant_id)
+    return devices, plant_id
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
```

Comment:
Let's extend the `ConfigEntry` to type the `entry.runtime_data` field. Example can be found here https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/runtime-data